File: web/src/utils/torusGenerator.ts
import { Complex } from './ellipticMath'
import { pyodideManager, TorusMeshData } from './pyodide'

export interface Vertex3D {
  x: number
  y: number
  z: number
}

export interface Facet {
  v1: number
  v2: number
  v3: number
  v4: number
}

export interface TorusGeometry {
  vertices: Vertex3D[]
  facets: Facet[]
  jInvariant: Complex
  discriminant: Complex
  tau: Complex
  usePython?: boolean
  metadata?: {
    p: number
    q: number
    degree: number
    meshDensity: number
    g2: string
    g3: string
    jInvariant: string
  }
}

/**
 * Generate lattice points for degree d approximation
 */
function generateLatticePoints(period1: Complex, period2: Complex, degree: number): Complex[] {
  const points: Complex[] = []
  const scale = 1.0 / Math.pow(2, degree) // 2^(-degree)
  
  // Generate points in a reasonable range for visualization
  const range = 10
  for (let n1 = -range; n1 <= range; n1++) {
    for (let n2 = -range; n2 <= range; n2++) {
      const point = period1.scale(n1 * scale).add(period2.scale(n2 * scale))
      points.push(point)
    }
  }
  return points
}

/**
 * Project complex lattice points to 3D torus surface using lattice-dependent parameters
 */
function projectToTorus(
  latticePoints: Complex[], 
  period1: Complex, 
  period2: Complex,
  meshDensity: number,
  p: number,
  q: number,
  degree: number
): Vertex3D[] {
  const vertices: Vertex3D[] = []
  
  // Make torus parameters dependent on p, q, and degree
  // The lattice L_d = Z(p * 2^(-d)) + Z(q * 2^(-d) * i) should affect the torus shape
  const latticeScale = Math.pow(2, -degree)  // 2^(-d)
  const majorRadius = 2.0 + 0.5 * Math.log(p + 1) * latticeScale
  const minorRadius = 0.5 + 0.2 * Math.log(q + 1) * latticeScale
  
  // Add aspect ratio based on p/q ratio
  const aspectRatio = Math.min(p, q) / Math.max(p, q)
  const radiusModulation = 1.0 + 0.3 * (1 - aspectRatio)
  
  for (let i = 0; i < meshDensity; i++) {
    for (let j = 0; j < meshDensity; j++) {
      const u = 2 * Math.PI * i / meshDensity
      const v = 2 * Math.PI * j / meshDensity
      
      // Add lattice-inspired modulations
      const pModulation = 0.1 * Math.sin(p * u) * latticeScale
      const qModulation = 0.1 * Math.cos(q * v) * latticeScale
      const degreeModulation = 0.05 * Math.sin(degree * (u + v))
      
      const effectiveMajor = majorRadius * radiusModulation + pModulation
      const effectiveMinor = minorRadius + qModulation
      
      const x = (effectiveMajor + effectiveMinor * Math.cos(v)) * Math.cos(u)
      const y = (effectiveMajor + effectiveMinor * Math.cos(v)) * Math.sin(u)
      const z = effectiveMinor * Math.sin(v) + degreeModulation
      
      vertices.push({ x, y, z })
    }
  }
  
  return vertices
}

/**
 * Generate quadrilateral facets for the mesh
 */
function generateFacets(meshDensity: number): Facet[] {
  const facets: Facet[] = []
  
  for (let i = 0; i < meshDensity; i++) {
    for (let j = 0; j < meshDensity; j++) {
      const current = i * meshDensity + j
      const nextI = ((i + 1) % meshDensity) * meshDensity + j
      const nextJ = i * meshDensity + (j + 1) % meshDensity
      const nextBoth = ((i + 1) % meshDensity) * meshDensity + (j + 1) % meshDensity
      
      facets.push({
        v1: current,
        v2: nextI,
        v3: nextBoth,
        v4: nextJ
      })
    }
  }
  
  return facets
}

/**
 * Main function to generate torus geometry using Python elliptic functions
 */
export async function generateTorusGeometry(
  p: number,
  q: number,
  degree: number,
  meshDensity: number = 20,
  usePython: boolean = true
): Promise<TorusGeometry> {
  // Try Python implementation first if requested and available
  if (usePython && pyodideManager.isReady()) {
    try {
      const pythonResult = await pyodideManager.generateTorusMesh(p, q, degree, meshDensity)
      
      // Calculate tau for compatibility
      const period1 = new Complex(p, 0)
      const period2 = new Complex(0, q)
      const tau = period2.divide(period1)
      
      return {
        vertices: pythonResult.vertices,
        facets: pythonResult.facets,
        jInvariant: new Complex(parseFloat(pythonResult.metadata.jInvariant) || 1728, 0),
        discriminant: period1.multiply(period2),
        tau,
        usePython: true,
        metadata: pythonResult.metadata
      }
    } catch (error) {
      console.warn('Python mesh generation failed, falling back to classical:', error)
    }
  }

  // Fallback to classical torus generation
  return generateClassicalTorusGeometry(p, q, degree, meshDensity)
}

/**
 * Generate classical torus geometry (original implementation with lattice-dependent parameters)
 */
export function generateClassicalTorusGeometry(
  p: number,
  q: number,
  degree: number,
  meshDensity: number = 20
): TorusGeometry {
  // Create lattice periods
  const period1 = new Complex(p, 0)
  const period2 = new Complex(0, q)
  
  // Calculate tau = period2 / period1
  const tau = period2.divide(period1)
  
  // Generate lattice points for degree d approximation
  const latticePoints = generateLatticePoints(period1, period2, degree)
  
  // Project to torus surface with lattice-dependent parameters
  const vertices = projectToTorus(latticePoints, period1, period2, meshDensity, p, q, degree)
  
  // Generate facets (quadrilaterals)
  const facets = generateFacets(meshDensity)
  
  // Calculate elliptic invariants (approximated based on lattice)
  const latticeScale = Math.pow(2, -degree)
  const g2Approx = 60 * Math.pow(p * latticeScale, -4) + 60 * Math.pow(q * latticeScale, -4)
  const g3Approx = 140 * Math.pow(p * latticeScale, -6) + 140 * Math.pow(q * latticeScale, -6)
  const discriminant = Math.pow(g2Approx, 3) - 27 * Math.pow(g3Approx, 2)
  const jInvariant = discriminant !== 0 ? new Complex(1728 * Math.pow(g2Approx, 3) / discriminant, 0) : new Complex(1728, 0)
  
  return {
    vertices,
    facets,
    jInvariant,
    discriminant: period1.multiply(period2),
    tau,
    usePython: false,
    metadata: {
      p,
      q,
      degree,
      meshDensity,
      g2: g2Approx.toFixed(6),
      g3: g3Approx.toFixed(6), 
      jInvariant: jInvariant.real.toFixed(6)
    }
  }
}

File: core/src/commonMain/kotlin/TorusGenerator.kt
package com.pqtorus.core

import kotlin.math.*

/**
 * Core class for generating elliptic curve tori visualization
 */
class TorusGenerator {
    
    /**
     * Generate torus geometry for given parameters
     * 
     * Creates elliptic curve torus visualization for lattice ℂ/L where:
     * L = {n₁p + n₂qi | n₁,n₂ ∈ ℤ} with prime periods p,q
     * 
     * Mathematical formulas:
     * - Lattice periods: ω₁ = p + 0i, ω₂ = 0 + qi
     * - Tau: τ = ω₂/ω₁ = qi/p = i(q/p)  
     * - Degree-d scaling: L_d = L/2^d
     * - J-invariant: 1728 (placeholder for j(τ))
     * - Discriminant: p*qi (simplified form)
     * 
     * @param p First prime period (real)
     * @param q Second prime period (real)  
     * @param degree Degree of approximation (d >= 0)
     * @param meshDensity Number of subdivisions for the mesh
     * @return TorusGeometry containing vertices, facets, and invariants
     */
    fun generateTorus(p: Double, q: Double, degree: Int, meshDensity: Int = 20): TorusGeometry {
        // Create lattice periods
        val period1 = Complex(p, 0.0)
        val period2 = Complex(0.0, q)
        
        // Calculate tau = period2 / period1
        val tau = if (period1.real != 0.0) {
            Complex(period2.real / period1.real, period2.imag / period1.real)
        } else {
            Complex(0.0, 1.0) // Default fallback
        }
        
        // Generate lattice points for degree d approximation
        val latticePoints = generateLatticePoints(period1, period2, degree)
        
        // Project to torus surface and create 3D vertices
        val vertices = projectToTorus(latticePoints, period1, period2, meshDensity)
        
        // Generate facets (quadrilaterals)
        val facets = generateFacets(meshDensity)
        
        // Calculate elliptic invariants
        val invariants = calculateEllipticInvariants(period1, period2)
        
        return TorusGeometry(
            vertices = vertices,
            facets = facets,
            jInvariant = invariants.jInvariant,
            discriminant = invariants.discriminant,
            tau = tau
        )
    }
    
    /**
     * Generate lattice points for degree d approximation
     * 
     * Formula: L_d = {(n₁ω₁ + n₂ω₂)/2^d | n₁,n₂ ∈ ℤ}
     * Where ω₁ = p, ω₂ = qi, and scale = 2^(-d)
     * 
     * @param period1 First lattice period (p + 0i)
     * @param period2 Second lattice period (0 + qi)
     * @param degree Approximation degree d ≥ 0
     * @return List of complex lattice points
     */
        val points = mutableListOf<Complex>()
        val scale = 1.0 / (1 shl degree) // 2^(-degree)
        
        // Generate points in the fundamental domain
        val range = 10 // Reasonable range for visualization
        for (n1 in -range..range) {
            for (n2 in -range..range) {
                val point = period1 * (n1 * scale) + period2 * (n2 * scale)
                points.add(point)
            }
        }
        return points
    }
    
    /**
     * Project lattice points to 3D torus surface
     * 
     * Current implementation uses standard geometric torus:
     * x(u,v) = (R + r*cos(v)) * cos(u)
     * y(u,v) = (R + r*cos(v)) * sin(u)  
     * z(u,v) = r * sin(v)
     * 
     * Where R = 2.0 (major radius), r = 0.5 (minor radius)
     * u,v ∈ [0,2π] parametrize the surface
     * 
     * Note: Actual lattice points not used in current projection.
     * Proper elliptic torus would use Weierstrass ℘-function.
     * 
     * @param latticePoints Generated lattice points (unused in current implementation)
     * @param period1 First lattice period
     * @param period2 Second lattice period  
     * @param meshDensity Grid resolution for torus mesh
     * @return List of 3D vertices
     */
        latticePoints: List<Complex>, 
        period1: Complex, 
        period2: Complex,
        meshDensity: Int
    ): List<Vertex3D> {
        val vertices = mutableListOf<Vertex3D>()
        
        // Create a regular torus parametrization
        val majorRadius = 2.0
        val minorRadius = 0.5
        
        for (i in 0 until meshDensity) {
            for (j in 0 until meshDensity) {
                val u = 2 * PI * i / meshDensity
                val v = 2 * PI * j / meshDensity
                
                val x = (majorRadius + minorRadius * cos(v)) * cos(u)
                val y = (majorRadius + minorRadius * cos(v)) * sin(u)
                val z = minorRadius * sin(v)
                
                vertices.add(Vertex3D(x, y, z))
            }
        }
        
        return vertices
    }
    
    private fun generateFacets(meshDensity: Int): List<Facet> {
        val facets = mutableListOf<Facet>()
        
        for (i in 0 until meshDensity) {
            for (j in 0 until meshDensity) {
                val current = i * meshDensity + j
                val nextI = ((i + 1) % meshDensity) * meshDensity + j
                val nextJ = i * meshDensity + (j + 1) % meshDensity
                val nextBoth = ((i + 1) % meshDensity) * meshDensity + (j + 1) % meshDensity
                
                facets.add(Facet(current, nextI, nextBoth, nextJ))
            }
        }
        
        return facets
    }
    
    private data class EllipticInvariants(
        val jInvariant: Complex,
        val discriminant: Complex
    )
    
    /**
     * Calculate elliptic curve invariants
     * 
     * Mathematical formulas for lattice with periods ω₁ = p, ω₂ = qi:
     * - Tau: τ = ω₂/ω₁ = qi/p = i(q/p)
     * - J-invariant: j(τ) = 1728*g₂³/(g₂³-27g₃²) [currently placeholder: 1728]
     * - Discriminant: Δ = Im(ω̄₁ω₂) = pq [currently simplified as p*qi]
     * 
     * Current implementation uses simplified calculations.
     * See MATHEMATICAL_FORMULAS.md for proper elliptic function theory.
     * 
     * @param period1 First lattice period (p + 0i)
     * @param period2 Second lattice period (0 + qi)
     * @return EllipticInvariants containing j-invariant and discriminant
     */
        // Simplified calculation for demonstration
        // In a full implementation, these would be proper elliptic function calculations
        
        // Calculate tau = period2 / period1
        val tau = period2 / period1
        
        // Simplified j-invariant approximation
        val jInvariant = Complex(1728.0, 0.0) // Placeholder
        
        // Simplified discriminant
        val discriminant = period1 * period2
        
        return EllipticInvariants(jInvariant, discriminant)
    }
}

/**
 * Check if a number is prime (simple implementation)
 */
fun isPrime(n: Int): Boolean {
    if (n < 2) return false
    if (n == 2) return true
    if (n % 2 == 0) return false
    
    for (i in 3..sqrt(n.toDouble()).toInt() step 2) {
        if (n % i == 0) return false
    }
    return true
}

/**
 * Get the next prime number >= n
 */
fun nextPrime(n: Int): Int {
    var candidate = maxOf(2, n)
    while (!isPrime(candidate)) {
        candidate++
    }
    return candidate
}